WIP: Amplitude integration

roman-mazur opened the pull request:
This is an initial pass to add Amplitude integration. Node side is a TODO item.
I'm currently testing the browser integration.

roman-mazur committed:
amplitude: Add browser implementation

Signed-off-by: Roman Mazur <roman@balena.io>

roman-mazur commented:
I wanted to open this to get some early feedback.

I also think what is the best way to support Amplitude `Identify` object with `append`, `setOnce`, etc operators.

DavidWells commented on packages/analytics-plugin-amplitude/src/browser.js:
```diff
@@ -0,0 +1,58 @@
+  let client = null, amplitudeInitCompleted = false;
+
+  return {
+    NAMESPACE: 'amplitude',
```
```suggestion
    name: 'amplitude',
```

DavidWells commented on packages/analytics-plugin-amplitude/src/node.js:
```diff
@@ -0,0 +1,18 @@
+ * })
+ */
+export default function amplitude() {
+
```
```suggestion
return {
  name: 'amplitude'
}
```

DavidWells commented:
Hey @roman-mazur 

Thanks for the PR it looks great!

It looks like segment uses an array of keys to tell which traits to only set once https://github.com/segmentio/analytics.js-integrations/blob/master/integrations/amplitude/lib/index.js#L208

Something like that could work.

Another thing to think about is initializing the library as a loaded script or exposing amplitude to the `window`. If it's available in the `window` people can use those "escape hatches" to call any of the raw methods from the sdk https://help.amplitude.com/hc/en-us/articles/115002889587-JavaScript-SDK-Reference

For example, segment is using the snippet from amplitude to load the SDK. https://github.com/segmentio/analytics.js-integrations/blob/master/integrations/amplitude/lib/index.js#L71 this exposes it to the window (which you could argue is bad or good ¯\_(ツ)_/¯). 

Loading from a script also keeps the apps bundle smaller by not inlining the entire library. For example it's possible for plugin `initialize` calls to be cancelled before they happen so amplitude wouldn't get loaded at all

roman-mazur commented:
>Loading from a script also keeps the apps bundle smaller by not inlining the entire library.

That's a good point. But in this case, we should also expose a plugin option that will allow changing where the script can be loaded from.

>If it's available in the window people can use those "escape hatches" to call any of the raw methods from the sdk.

Having amplitude exposed as a global variable could help with existing code base migration.
However, for escape hatches, would not this be better to get a reference to the client instance trough the plugin? Something like
```js
const amplitudeIntegration = amplitude({
  apiKey: '...',
  projectName: '...'
})
const analytics = new Analytics({
  plugins: [amplitudeIntegration]
})

amplitudeIntegration.amplitude.identify();
```

DavidWells commented:
> That's a good point. But in this case, we should also expose a plugin option that will allow changing where the script can be loaded from.

I think including the library is fine. 

> Having amplitude exposed as a global variable could help with existing code base migration.
However, for escape hatches, would not this be better to get a reference to the client instance trough the plugin? 

I like this idea. We will need to figure out an API for plugins to expose custom methods like this. I've been thinking about this recently. Something like:

```js
export default function plugin(userConfig) {
  return {
    name: 'my-example-plugin',
    config: {},
    // Methods are custom functions added to analytics instance
    methods: {
       thing: () => { /* custom functionality */ },
       otherThing: () => { /* custom functionality */ },
    },
    initialize: ({ config }) => {},
    page: ({ payload }) => {},
  }
}
```

Then it would be possible to call `thing` and `otherThing` with the `analytics` instance like:

```js
analytics.thing()
analytics.otherThing()
```

---

The PR looks great though. Let me know when its good to merge and we can get it published to `npm` 😃

roman-mazur commented:
@DavidWells sorry for the delay

I've finally finished testing the basic browser integration and now working on nodejs.
This browser implementation should be good to merge.

roman-mazur committed:
Provide an escape hatch for Amplitude

Use like
    const amplPlugin = amplitude(config);
    const analytics = new Analytics(plugins: [amplPlugin]);
    amplPlugiin.amplitude.getInstance().identify(new Identify());

Signed-off-by: Roman Mazur <roman@balena.io>

roman-mazur committed:
Add node implementation

Signed-off-by: Roman Mazur <roman@balena.io>

roman-mazur committed:
Bump amplitude library version

Signed-off-by: Roman Mazur <roman@balena.io>

roman-mazur commented:
After playing with this for a while, I ended up implementing 2 separate modules for browser and node: dependencies (and implementations) are too different.
I have tried to make it a single module, but ran into multiple issues importing this package and using it both in browser and in nodejs. But I'm open to suggestions.

thilak-rao commented:
@roman-mazur Great work on this merge request. Let me know if there is anything I could help you with?

johnparn commented:
Hi!

What is the state of this integration - is the Amplitude integration about to be merged anytime soon? The reason for asking is that I may have to use Amplitude along with GA in a project and it would be really nice to be able to use analytics.

roman-mazur commented:
Closing in favor of #165
